src/main/scala/model/entities/Bullet.scala

```diff
@@ -47,28 +47,40 @@ abstract class Bullet(position: Position) extends Entity with MovingAbility:
  * The [[Bullet]] shoot by the [[Peashooter]]. It has no strange effects on the zombies beside dealing damage.
  * @param position The initial position of the [[Bullet]].
  */
-case class PeaBullet(override val position: Position = (0,0)) extends Bullet(position):
+case class PeaBullet(override val position: Position) extends Bullet(position):
   override def checkCollisionWith(entity: Entity): Boolean =
     entity match
       case _: Enemy => super.checkCollisionWith(entity)
       case _ => false
 
   override def update(elapsedTime: FiniteDuration, interests: List[Entity]): Bullet =
-    PeaBullet() withPosition newPositionAfter(elapsedTime)
+    this withPosition newPositionAfter(elapsedTime)
 
   override def withPosition(pos: Position): Bullet = copy(position = pos)
 
+
 /**
- * The [[Bullet]] shoot by the [[Zombie]]. It has no strange effects on the plants beside dealing damage.
+ * An abstract [[Entity]] shoot by a [[Zombie]].
  * @param position The initial position of the [[Bullet]].
  */
-case class PawBullet(override val position: Position = (0,0)) extends Bullet(position):
-  override def update(elapsedTime: FiniteDuration, interests: List[Entity]): Bullet =
-    PawBullet() withPosition newPositionAfter(elapsedTime)
-
+abstract class ZombieAttack(override val position: Position) extends Bullet(position):
   override def checkCollisionWith(entity: Entity): Boolean =
     entity match
       case _: Plant => super.checkCollisionWith(entity)
       case _ => false
 
-  override def withPosition(pos: Position): Bullet = copy(position = pos)
\ No newline at end of file
+/**
+ * The [[ZombieAttack]] shoot by the classic[[Zombie]] and [[FastZombie]]. It has no strange effects on the plants beside dealing damage.
+ * @param position The initial position of the [[Bullet]].
+ */
+case class PawBullet(override val position: Position) extends ZombieAttack(position):
+  override def update(elapsedTime: FiniteDuration, interests: List[Entity]): Bullet = this withPosition newPositionAfter(elapsedTime)
+  override def withPosition(pos: Position): Bullet = copy(position = pos)
+
+/**
+ * The [[ZombieAttack]] shoot by  [[WarriorZombie]]. It has no strange effects on the plants beside dealing damage.
+ * @param position The initial position of the [[Bullet]].
+ */
+case class SwordAttack(override val position: Position) extends ZombieAttack(position):
```

Comment:
Every entity shoot should be a bullet.
A zombie shoots a SwordBullet not a SwordAttack.
Change the name and use the concept of Bullet instead of the concept of Attack.
The same goes for the abstract class ZombieAttack (should be ZombieBullet)